### PR TITLE
PI-1478 changes to retry exceptions

### PR DESCRIPTION
**Pull request by stevomcallister**

### CODE REVIEW

File: libs/messaging/src/main/kotlin/uk/gov/justice/digital/hmpps/listener/AwsNotificationListener.kt [hunk 1]
```diff
@@ -25,7 +31,16 @@ class AwsNotificationListener(
...
         try {
-            handler.handle(message)
+            retry(
+                3,
+                listOf(
+                    FeignException.NotFound::class,
+                    RetryableException::class,
+                    ObjectOptimisticLockingFailureException::class,
+                    CannotCreateTransactionException::class,
+                    CannotGetJdbcConnectionException::class
```
Comment (marcus-bcl) [inline libs/messaging/src/main/kotlin/uk/gov/justice/digital/hmpps/listener/AwsNotificationListener.kt:41]: Is it worth including `CannotAcquireLockException` here too?

File: libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt [hunk 2]
```diff
@@ -1,13 +1,24 @@
 package uk.gov.justice.digital.hmpps.retry
 
-fun <T> retry(maxRetries: Int, code: () -> T): T {
-    var throwable: Throwable? = null
-    (1..maxRetries).forEach { _ ->
+import kotlin.reflect.KClass
+
+fun <T> retry(maxRetries: Int, exceptions: List<KClass<out Exception>> = listOf(Exception::class), code: () -> T): T {
+    var throwable: Throwable?
+    (1..maxRetries).forEach { count ->
```
Comment (marcus-bcl) [inline libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt:7]: Do we need any kind of delay or back-off between retries?

File: libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+            val matchedException = exceptions.firstOrNull { it.isInstance(e) }
+            throwable = if (matchedException != null && count < maxRetries) {
+                null
+            } else {
+                e
+            }
+
+            if (throwable != null) {
+                throw throwable!!
+            }
```
Comment (marcus-bcl) [inline libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt:20]: I'm a bit confused by this... probably just me... where do we return the successful value outside the forEach?

File: libs/messaging/src/main/kotlin/uk/gov/justice/digital/hmpps/listener/AwsNotificationListener.kt (hunk 1 above)
Comment (stevomcallister) [reply libs/messaging/src/main/kotlin/uk/gov/justice/digital/hmpps/listener/AwsNotificationListener.kt:41]: ok

File: libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt (hunk 2 above)
Comment (stevomcallister) [reply libs/commons/src/main/kotlin/uk/gov/justice/digital/hmpps/retry/Retry.kt:7]: added a delay and using Feign retry stuff

**Pull request merged**